File: backend/app/routers/wechat.py
# -*- coding: utf-8 -*-
"""微信 iLink 登录和连接状态接口。"""
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel, Field

from ..wechat.ilink_client import ILinkError
from ..wechat.service import wechat_service

router = APIRouter(prefix='/api/wechat')


class WeChatPolicyBody(BaseModel):
    allow_all: bool = False
    allow_users: list[str] = Field(default_factory=list, max_length=200)


@router.get('/status')
def wechat_status():
    return wechat_service.status()


@router.get('/config')
def wechat_config():
    return wechat_service.policy()


@router.put('/config')
def update_wechat_config(body: WeChatPolicyBody):
    users = []
    for user_id in body.allow_users:
        normalized = user_id.strip()
        if normalized and normalized not in users:
            users.append(normalized)
    return {'ok': True, **wechat_service.update_policy(users, body.allow_all)}


@router.post('/login/start')
async def start_login():
    try:
        return await wechat_service.start_login()
    except ILinkError as exc:
        raise HTTPException(502, str(exc)) from exc


@router.post('/login/poll')
async def poll_login():
    try:
        return await wechat_service.poll_login()
    except ILinkError as exc:
        raise HTTPException(502, str(exc)) from exc


@router.post('/loop/start')
async def start_loop():
    try:
        return await wechat_service.start_loop()
    except ILinkError as exc:
        raise HTTPException(409, str(exc)) from exc


@router.post('/loop/stop')
async def stop_loop():
    await wechat_service.stop()
    return {'running': False}


@router.post('/reminders/send')
async def send_reminders():
    try:
        return await wechat_service.send_pending_reminders()
    except Exception as exc:
        raise HTTPException(400, f'微信提醒发送失败：{exc}') from exc

File: server/src/http/routes/wechat.ts
import type { FastifyInstance, FastifyReply } from 'fastify';

import { ILinkError } from '../../wechat/ilinkClient.js';
import { wechatService } from '../../wechat/service.js';

function mapError(reply: FastifyReply, status: number, error: unknown): FastifyReply | undefined {
  if (error instanceof ILinkError) {
    return reply.status(status).send({ detail: error.message });
  }
  return undefined;
}

export function registerWechatRoutes(app: FastifyInstance): void {
  app.get('/api/wechat/status', async () => wechatService.status());

  app.get('/api/wechat/config', async () => wechatService.policy());

  app.put('/api/wechat/config', async (request, reply) => {
    const body = (request.body ?? {}) as { allow_all?: boolean; allow_users?: string[] };
    if (Array.isArray(body.allow_users) && body.allow_users.length > 200) {
      return reply.status(422).send({ detail: '请求参数校验失败' });
    }
    const users: string[] = [];
    for (const userId of Array.isArray(body.allow_users) ? body.allow_users : []) {
      const normalized = String(userId).trim();
      if (normalized && !users.includes(normalized)) users.push(normalized);
    }
    return { ok: true, ...wechatService.updatePolicy(users, body.allow_all === true) };
  });

  app.post('/api/wechat/login/start', async (_request, reply) => {
    try {
      return await wechatService.startLogin();
    } catch (error) {
      const mapped = mapError(reply, 502, error);
      if (mapped) return mapped;
      throw error;
    }
  });

  app.post('/api/wechat/login/poll', async (_request, reply) => {
    try {
      return await wechatService.pollLogin();
    } catch (error) {
      const mapped = mapError(reply, 502, error);
      if (mapped) return mapped;
      throw error;
    }
  });

  app.post('/api/wechat/loop/start', async (_request, reply) => {
    try {
      return await wechatService.startLoop();
    } catch (error) {
      const mapped = mapError(reply, 409, error);
      if (mapped) return mapped;
      throw error;
    }
  });

  app.post('/api/wechat/loop/stop', async () => {
    await wechatService.stop();
    return { running: false };
  });

  app.post('/api/wechat/reminders/send', async (_request, reply) => {
    try {
      return await wechatService.sendPendingReminders();
    } catch (error) {
      return reply.status(400).send({ detail: `微信提醒发送失败：${(error as Error).message}` });
    }
  });
}
